query.go

```diff
@@ -51,16 +60,47 @@ func GetQuery(query backend.DataQuery, headers http.Header, setHeaders bool) (*Q
 	return model, nil
 }
 
-// QueryDB sends the query to the connection and converts the rows to a dataframe.
-func QueryDB(ctx context.Context, db Connection, converters []sqlutil.Converter, fillMode *data.FillMissing, query *Query, args ...interface{}) (data.Frames, error) {
+type DBQuery struct {
+	DSName     string
+	DB         Connection
+	Settings   backend.DataSourceInstanceSettings
+	converters []sqlutil.Converter
+	fillMode   *data.FillMissing
+}
+
+func NewQuery(db Connection, settings backend.DataSourceInstanceSettings, converters []sqlutil.Converter, fillMode *data.FillMissing) *DBQuery {
+	dsName, ok := sanitizeLabelName(settings.Name)
+	if !ok {
+		backend.Logger.Warn("Failed to sanitize datasource name for prometheus label", dsName)
+	}
+	return &DBQuery{
+		DB:         db,
+		DSName:     dsName,
+		converters: converters,
+		fillMode:   fillMode,
+	}
+}
+
+// Run sends the query to the connection and converts the rows to a dataframe.
+func (q *DBQuery) Run(ctx context.Context, query *Query, args ...interface{}) (data.Frames, error) {
 	// Query the rows from the database
-	rows, err := db.QueryContext(ctx, query.RawSQL, args...)
+	start := time.Now()
+	var errorSource = "none"
+
+	defer func() {
+		if q.DSName != "" {
+			duration.WithLabelValues(q.DSName, q.Settings.Type, errorSource).Observe(time.Since(start).Seconds())
+		}
+	}()
+
+	rows, err := q.DB.QueryContext(ctx, query.RawSQL, args...)
 	if err != nil {
 		errType := ErrorQuery
 		if errors.Is(err, context.Canceled) {
 			errType = context.Canceled
 		}
 		err := DownstreamError(fmt.Errorf("%w: %s", errType, err.Error()))
+		errorSource = string(backend.ErrorSourceDownstream)
```

Comment:
Don't understand why do we need `errorSource` separately? Can't we get the error source from the error itself?

Reply:
updated